subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DependenciesExtensionModule.java

```diff
@@ -239,22 +242,83 @@ public static <D extends Dependency> void call(DependencyCollector self, D depen
         self.add(dependency, ConfigureUtil.configureUsing(configuration));
     }
 
+    @VisibleForTesting
+    public static final String ERROR_MESSAGE_PROVIDER = "Providers of non-Dependency(Constraint) types (java.lang.String) are not supported. Create a Dependency(Constraint) using Dependency(Constraint)Factory first.";
```

Comment:
❌ This message isn't very easy to understand. I see that you're trying to use `(Constraint)` to be a sort of optional part of the message, but that gets in the way.

I would twist this around to make the type that's provided first and say what's supported later, something like:
> Providers of type '%s' are not supported. Only Provider<Dependency> or Provider<DependencyConstraint> are allowed. 

We could link to some docs if we think that would be helpful too.

Reply:
Refactored out.